1582479 - Document Glean crash capabilities as a use-case focus (#399)
mozilla/glean: docs/user/instrument-android-crashes-example.md

override of the `uncaughtException()` function.  An example of this override that records data and sends the ping could look
something like this:

```Kotlin
override fun uncaughtException(thread: Thread, exception: Throwable) {
    Crash.exception.record(
        mapOf(
            Crash.exceptionKeys.cause to exception.cause!!.toString(),
            Crash.exceptionKeys.message to exception.message!!)
    )
    Pings.crash.send()
}
```

This records data to the `Crash.exception` metric from the `metrics.yaml`.  The category of the metric is `crash` and the name
is `exception` so it is accessed it by calling `record()` on the `Crash.exception` object.  The extra information for the
`cause` and the `message` is set as well.  Finally, calling `Pings.crash.send()` forces the `crash` ping to be scheduled to be
sent.

The final step is to register the custom `Application` as the default uncaught exception handler by adding the following to the
`onCreate()` function after `Glean.initialize(this)`:

```Kotlin
Thread.setDefaultUncaughtExceptionHandler(this)
```

### Next Steps

> This information didn't really get recorded by anything, as it would be rejected by the telemetry pipeline unless the

Reviewer: What do you mean here? Can you please expand on what's missing?

Author: I added the same link to the `probe_scraper` docs as found in the "Adding Glean to your project" section, which should cover the "next steps" required.

### Add A Custom Ping

Define the custom ping that will help control the upload scheduling by creating a `pings.yaml` file in the same directory as
the `metrics.yaml` file.  For more information about adding custom pings, see the section on [custom pings](pings/custom.md).  
The name of the ping will be `crash`, so the `pings.yaml` file should look like this:

```YAML
# Required to indicate this is a `pings.yaml` file
$schema: moz://mozilla.org/schemas/glean/pings/1-0-0

crash:
  description: >
    A ping to transport crash data
  include_client_id: true
  notification_emails:
    - crash@example.com
  bugs:
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
  data_reviews:
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
```

Before the newly defined metric or ping can be used, the application must first be built.  This will cause the [glean_parser](https://github.com/mozilla/glean_parser/)
to execute and generate the API files that represent the metric and ping that were newly defined.

> **NOTE:** If changes to the YAML files aren't showing up in the project, try running the clean task on the project before

Reviewer 2: This document has "Important Note", "Note" and "NOTE". We should probably make this a consistent "Note"

building any time one of the Glean YAML files has been modified.  

It is recommended that Glean be initialized as early in the application startup as possible, which is why it's good to use a

> custom `Application`, like the Glean Sample App `GleanApplication` found [here](https://github.com/mozilla/glean/blob/master/samples/android/app/src/main/java/org/mozilla/samples/glean/GleanApplication.kt)

Reviewer 2: "here" is a bad click target (small and common word).
Use something like "found in the [Glean repository]" or link the longer "GleanApplication" directly

override fun onCreate() {
  super.onCreate()

  // Register the application's custom pings.
  Glean.registerPings(Pings)

  // Initialize the Glean library
  Glean.initialize(applicationContext)
}

```

This completes the registration of the custom ping with the Glean SDK so that it knows about it and can manage the storage and
other important details of it like sending it when `send()` is called.

### Instrument The App To Record The Event

In order to make the custom `Application` class handle uncaught exceptions, extend the class definition by adding
`Thread.UncaughtExceptionHandler` as an inherited class like this:

```Kotlin
class MainActivity : AppCompatActivity(), Thread.UncaughtExceptionHandler {
    ...
}
```

As part of implementing the `Thread.UncaughtExceptionHandler` interface, the custom `Application` needs to implement the
override of the `uncaughtException()` function.  An example of this override that records data and sends the ping could look
something like this:

```Kotlin
override fun uncaughtException(thread: Thread, exception: Throwable) {
    Crash.exception.record(
        mapOf(
            Crash.exceptionKeys.cause to exception.cause!!.toString(),

>             Crash.exceptionKeys.message to exception.message!!)

Reviewer 2: *ktlint mode* The `)` should probably be on the next line, aligned with `mapOf`

> # Recording Android Crashes with Glean

Reviewer 3: This uses the word "instrumenting" rather than "recording" every where this is linked to.  I guess I prefer "instrumenting", but don't have a strong preference.  Let's make it consistent, though?

> For the last year I’ve been working on this shiny new library by Mozilla known as [Glean SDK](https://github.com/mozilla/glean). The Glean SDK is a cross-platform telemetry library that makes collecting data easier for developers to understand and instrument in their applications.  Just take a look to the tagline, “Telemetry for humans” to understand the objective of Glean.  Glean is being used right now in things like [Firefox Preview](https://github.com/mozilla-mobile/fenix/) and [Firefox for Fire TV](https://github.com/mozilla-mobile/firefox-tv/) and I’m sure we will see it in even more Mozilla products very soon.

Reviewer 3: This paragraph is great for a blog post, and I don't mind the friendly tone in docs at all, but it does seem a bit redundant to the description of Glean elsewhere in these docs.  Also, I worry that it will get out of date real fast (whereas a blog post, like an article, is a "snapshot in time").

Author: I can save this as window dressing for the blog post.  It felt a little out of place to me in the docs as well.

> One of the things that we might want to collect some data on in the applications we create is crashes.  Every application will have them from time to time.  Sometimes they are the application’s fault, sometimes it’s the fault of the OS, but they happen and we usually want to know more about them so we can fix them (if possible).  So let’s take a look at how we can use Glean to instrument an Android application and record some data when it crashes.

Reviewer 3: Maybe semantic line breaks?

## Before We Start

> There are a few things that we need to have installed in order to get ready.  The biggest one of those is [Android Studio](https://developer.android.com/studio/), which (if you include the Android SDK) can take a little while to download and get installed.  Second, I’m going to recommend having the [Book of Glean](https://mozilla.github.io/glean/book/) open and available as I will be referring to it throughout this walk-through as it contains all the Glean documentation we need to instrument our app.  I’m also assuming that you have some knowledge of Android application development, we won’t be doing anything too involved, but knowing where to go to create a new project, and how to add dependencies to a Gradle file will be helpful.

Reviewer 3: ```suggestion
There are a few things that we need to have installed in order to get ready.  The biggest one of those is [Android Studio](https://developer.android.com/studio/), which (if you include the Android SDK) can take a little while to download and get installed.  Second, I’m going to recommend having the [Book of Glean](https://mozilla.github.io/glean/book/) open and available as I will be referring to it throughout this walk-through as it contains all the Glean documentation we need to instrument our app.  I’m also assuming that you have some knowledge of Android application development. We won’t be doing anything too involved, but knowing where to go to create a new project, and how to add dependencies to a Gradle file will be helpful.
```

## Create A Project

With our tools and documentation installed and open, let’s start by creating a new Android Studio project.  It could be any type of project to start with, but I’m choosing the “Basic Activity” template simply because it comes with a button already on the screen that can use to crash our app in order to collect our metrics.  Give the project a name of "GleanCrashExample", choose Kotlin for the language, and choose API level 21 for the minimum SDK level.

## Setup Build Configuration

### Add Glean As A Dependency

Now that we have a project started, we can add Glean to it as a dependency in the Gradle configuration.  Find the `build.gradle` file for your _app module_ and add the following line to the dependencies section:

```Groovy
implementation 'org.mozilla.components:service-glean:16.0.0'
```

> As of this writing, version `16.0.0` is the latest version of [Android-Components](https://github.com/mozilla-mobile/android-components/).  Android-Components is a collection of tools for building browsers on Android and it is the easiest way to consume Glean on Android at the moment.  You should probably take a look at the Android-Components releases and check that you are using the latest version of Android-Components for your project.  So just replace the version above with the latest.

Reviewer 3: The a-c docs seem to use "Android Components" or `android-components`, but never "Android-Components".  Maybe remove the hyphen to be consistent with them?

### Add Mozilla Maven Repository

> In order to be able to download Glean, we need to add Mozilla’s maven repository URL to the Project build.gradle file.  In Android Studio, for our project named “GleanCrashExample”, there will be a build.gradle that has `(Project: GleanCrashExample)` next to it.  In this Gradle file, you will find a section called `allProjects` which has in it a section called `repositories`.  You may already find `google()` and `jcenter()` there, but we need to add the following in order to download Glean:

Reviewer 3: ```suggestion
In order to be able to download Glean, we need to add Mozilla’s maven repository URL to the Project `build.gradle` file.  In Android Studio, for our project named “GleanCrashExample”, there will be a `build.gradle` that has `(Project: GleanCrashExample)` next to it.  In this Gradle file, you will find a section called `allProjects` which has in it a section called `repositories`.  You may already find `google()` and `jcenter()` there, but we need to add the following in order to download Glean:
```

```Groovy
maven {
    url "https://maven.mozilla.org/maven2"
}
```

### Add Python Environment

We add a Python environment for the [glean_parser](https://github.com/mozilla/glean_parser/) to run in with the following lines, added to the very top of the `build.gradle` file:

```Groovy
plugins {
    id "com.jetbrains.python.envs" version "0.0.26"
}
```

### Import SDK Generator Script

> The last part of the changes to the app’s build.gradle file is to include the `sdk_generator.gradle` script.  Right before the end of the `build.gradle` file, add the following (making sure the version here matches the version of the `implementation` line above):

Reviewer 3: ```suggestion
The last part of the changes to the app’s `build.gradle` file is to include the `sdk_generator.gradle` script.  Right before the end of the `build.gradle` file, add the following (making sure the version here matches the version of the `implementation` line above):
```

Glean.initialize(this)
```

That’s it!  The app now has Glean installed and enabled and will be able to send the baseline ping without any additional work!

### Add Custom Metric

Since we are instrumenting crashes with some custom metrics, we still have some work to do.  We will need to add a `metrics.yaml` file to define the metrics we will use to record our crash information and a `pings.yaml` file to define a custom ping which will give us some control over the scheduling of the uploading of the crash telemetry.

In order to do this we need to make a decision.  Mainly, what metric type will we use to represent our crash data.  This could probably be argued several ways, but I have chosen to instrument it as an [event](https://mozilla.github.io/glean/book/user/metrics/event.html), simply because events capture information in a nice concise way and have a built-in way of passing additional information using the `extras` field.  So if we want to pass along the cause, as well as a few lines of description or even potentially a few lines of stack trace, events let us do that easily (with [some limitations](https://mozilla.github.io/glean/book/user/metrics/event.html#limits)).

> Now that we have decided what metric type we will use, we can now create our `metrics.yaml`.  Inside of the `app` folder in the Android Studio project we are going to create a new file and call it `metrics.yaml`.  Then we can add the schema definition and the metric definition to the file, so that it looks like this:

Reviewer 3: ```suggestion
Now that we have decided what metric type we will use, we can create our `metrics.yaml`.  Inside of the `app` folder in the Android Studio project we are going to create a new file and call it `metrics.yaml`.  Then we can add the schema definition and the metric definition to the file, so that it looks like this:
```

$schema: moz://mozilla.org/schemas/glean/metrics/1-0-0

crash:
  exception:
    type: event
    description: |
      Event to record crashes caused by unhandled exceptions
    notification_emails:
      - crashes@example.com
    bugs:
      - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
    data_reviews:
      - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
    expires:
      2099-01-01
    send_in_pings:
      - crash
    extra_keys:
      cause:
        description: The cause of the crash
      message:
        description: The exception message
```

By way of explanation, this creates a metric called `exception` within a metric category called `crash`.  There is a brief description and the required notification, bug, data review, and expiry fields.  Then we have the `send_in_pings` field with a value of `- crash`.  This means that we will send the crash data via a custom ping named `crash` (which we haven’t created, yet).  Finally we have the `extra_keys` field with two keys defined, `cause` and `message`.  This will allow us to send a couple of pieces of additional information along with the event.

### Add Custom Ping

> Now we need to define the custom ping by creating a pings.yaml file in the same directory as we just created the `metrics.yaml` file.  We already know what the name of the ping is, `crash`, so the `pings.yaml` file should look like this:

Reviewer 3: ```suggestion
Now we need to define the custom ping by creating a `pings.yaml` file in the same directory as we just created the `metrics.yaml` file.  We already know what the name of the ping is, `crash`, so the `pings.yaml` file should look like this:
```

```YAML
# Required to indicate this is a `pings.yaml` file
$schema: moz://mozilla.org/schemas/glean/pings/1-0-0

crash:
  description: >
    A ping to transport crash data
  include_client_id: true
  notification_emails:
    - crash@example.com
  bugs:
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
  data_reviews:
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
```

Before we can use our newly defined metric or ping, we need to build the application.  This will cause the [glean_parser](https://github.com/mozilla/glean_parser/) to work its magic and generate the API files that represent the metrics we defined.

> **NOTE:** Just a little advice from my personal experience here, but I would recommend doing a Clean on your project before building _any_ time you have modified one of the Glean YAML files.  Sometimes it doesn't regenerate the files without the Clean so just something to keep in mind if you aren’t seeing your changes to the YAML files show up in the project.

Reviewer 3: "doing a Clean" -> maybe "running the `clean` task"

> We now have a couple of tasks to perform back in the MainActivity.kt file in order to make use of the metric and ping we defined.  First, we need to add an import line:

Reviewer 3: ```suggestion
We now have a couple of tasks to perform back in the `MainActivity.kt` file in order to make use of the metric and ping we defined.  First, we need to add an import line:
```

```Kotlin
import org.mozilla.gleancrashexample.GleanMetrics.Pings
```

> Then we need to register our custom ping by calling `Glean.initialize(this)`.  In the `MainActivity.onCreate()` function, add the following line:

Reviewer 3: ```suggestion
Then we need to register our custom ping by calling `Glean.registerPings(Pings)`.  In the `MainActivity.onCreate()` function, add the following line:
```

# Instrumenting Android Crashes with Glean

One of the things that we might want to collect some data on in the applications we create is crashes.  Every application will
have them from time to time.  Sometimes they are the application’s fault, sometimes it’s the fault of the OS, but they happen
and we usually want to know more about them so we can fix them (if possible).  So let’s take a look at how we can use Glean to
instrument an Android application and record some data when it crashes.

## Before We Start

There are a few things that we need to have installed in order to get ready.  The biggest one of those is [Android Studio](https://developer.android.com/studio/),
which (if you include the Android SDK) can take a little while to download and get installed.  Second, I’m going to recommend

> having the [Book of Glean](https://mozilla.github.io/glean/book/) open and available as we will be referring to it throughout

Reviewer 2: We refer to it as "The Glean SDK Book" in the readme. We did discuss the naming a while ago but I forgot what the decision was.

Reviewer 2: also given that this chapter is _inside_ that book we might want to rephrase this?

this walk-through as it contains all the Glean documentation we need to instrument our app.  This also assumes that you have
some knowledge of Android application development.  We won’t be doing anything too involved, but knowing where to go to
create a new project, and how to add dependencies to a Gradle file will be helpful.

## Create A Project

With our tools and documentation installed and open, let’s start by creating a new Android Studio project.  It could be any
type of project to start with, but I’m choosing the “Basic Activity” template simply because it comes with a button already on
the screen that can use to crash our app in order to collect our metrics.  Give the project a name of "GleanCrashExample",
choose Kotlin for the language, and choose API level 21 for the minimum SDK level.

## Setup Build Configuration

### Add Glean As A Dependency

Now that we have a project started, we can add Glean to it as a dependency in the Gradle configuration.  Find the
`build.gradle` file for your _app module_ and add the following line to the dependencies section:

```Groovy

> implementation 'org.mozilla.components:service-glean:16.0.0'

Reviewer 2: 18.0.0 is current, 19.0.0 is what will contain glean-core

```

> As of this writing, version `16.0.0` is the latest version of [Android Components](https://github.com/mozilla-mobile/android-components/).  

Reviewer 2: Maybe change this to state where the current version can be looked up.
This part will _always_ be out of date.

Android Components is a collection of tools for building browsers on Android and it is the easiest way to consume Glean on

> Android at the moment.  You should probably take a look at the Android-Components releases and check that you are using the

Reviewer 2: Ah, here it is. I would lead with that.

latest version of Android-Components for your project.  So just replace the version above with the latest.

### Add Mozilla Maven Repository

> In order to be able to download Glean, we need to add Mozilla’s maven repository URL to the Project `build.gradle` file.  In

Reviewer 2: ```suggestion
In order to be able to download Glean, we need to add Mozilla’s maven repository URL to the project's `build.gradle` file.  In
```

Android Studio, for our project named “GleanCrashExample”, there will be a `build.gradle` that has `(Project: GleanCrashExample)`
next to it.  In this Gradle file, you will find a section called `allProjects` which has in it a section called `repositories`.
You may already find `google()` and `jcenter()` there, but we need to add the following in order to download Glean:

```Groovy
maven {
    url "https://maven.mozilla.org/maven2"
}
```

### Add Python Environment

We add a Python environment for the [glean_parser](https://github.com/mozilla/glean_parser/) to run in with the following lines,
added to the very top of the `build.gradle` file:

```Groovy
plugins {
    id "com.jetbrains.python.envs" version "0.0.26"
}
```

### Import SDK Generator Script

The last part of the changes to the app’s `build.gradle` file is to include the `sdk_generator.gradle` script.  Right before the
end of the `build.gradle` file, add the following (making sure the version here matches the version of the `implementation`
line above):

```Groovy

> apply from: 'https://github.com/mozilla-mobile/android-components/raw/v16.0.0/components/service/glean/scripts/sdk_generator.gradle'

Reviewer 2: Should then be the version from above if you change that.

```

That should be it for additions to the build configuration files.  You should try to do a Gradle sync now to make sure

> everything works.  If you have any problems, try going back through the configuration steps to make sure you got everything.

Reviewer 2: As this will become part of the whole book, would it make sense to link back to https://mozilla.github.io/glean/book/user/adding-glean-to-your-project.html as a "short summary"?

Reviewer 2: (still keeping the section here of course)

## Instrument The Project With Glean

### Initial Integration

Now that we have added Glean to our app, we need to do a little initialization in order to get it ready to record and send
data. At the top of your `MainActivity.kt` file, add the following import statement:

```Kotlin
import mozilla.components.service.glean.Glean
```

Then inside of the `onCreate()` function of the `MainActivity`, add the following 2 lines to enable and initialize Glean:

```Kotlin
Glean.setUploadEnabled(true)
Glean.initialize(this)
```

That’s it!  The app now has Glean installed and enabled and will be able to send the baseline ping without any additional work!

> ### Add Custom Metric

Reviewer 2: ```suggestion
### Add a custom metric
```

Since we are instrumenting crashes with some custom metrics, we still have some work to do.  We will need to add a
`metrics.yaml` file to define the metrics we will use to record our crash information and a `pings.yaml` file to define a
custom ping which will give us some control over the scheduling of the uploading of the crash telemetry.

In order to do this we need to make a decision.  Mainly, what metric type will we use to represent our crash data.  This could

> probably be argued several ways, but I have chosen to instrument it as an [event](https://mozilla.github.io/glean/book/user/metrics/event.html),

Reviewer 2: Make it a relative link to the .md file and our linkchecker will be happy to ensure it keeps working

simply because events capture information in a nice concise way and have a built-in way of passing additional information
using the `extras` field.  So if we want to pass along the cause, as well as a few lines of description or even potentially a

> few lines of stack trace, events let us do that easily (with [some limitations](https://mozilla.github.io/glean/book/user/metrics/event.html#limits)).

Reviewer 2: same here

the metric definition to the file, so that it looks like this:

```YAML
# Required to indicate this is a `metrics.yaml` file
$schema: moz://mozilla.org/schemas/glean/metrics/1-0-0

crash:
  exception:
    type: event
    description: |
      Event to record crashes caused by unhandled exceptions
    notification_emails:
      - crashes@example.com
    bugs:
      - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
    data_reviews:
      - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
    expires:
      2099-01-01
    send_in_pings:
      - crash
    extra_keys:
      cause:
        description: The cause of the crash
      message:
        description: The exception message
```

By way of explanation, this creates a metric called `exception` within a metric category called `crash`.  There is a brief
description and the required notification, bug, data review, and expiry fields.  Then we have the `send_in_pings` field with a
value of `- crash`.  This means that we will send the crash data via a custom ping named `crash` (which we haven’t created,
yet).  Finally we have the `extra_keys` field with two keys defined, `cause` and `message`.  This will allow us to send a

> couple of pieces of additional information along with the event.

Reviewer 2: Worth mentioning that we require data review for all Mozilla applications.
See **Important** box on https://mozilla.github.io/glean/book/user/adding-new-metrics.html

> ### Add Custom Ping

Reviewer 2: ```suggestion
### Add a custom ping
```

Now we need to define the custom ping by creating a `pings.yaml` file in the same directory as we just created the `metrics.yaml`
file.  We already know what the name of the ping is, `crash`, so the `pings.yaml` file should look like this:

```YAML
# Required to indicate this is a `pings.yaml` file
$schema: moz://mozilla.org/schemas/glean/pings/1-0-0

crash:
  description: >
    A ping to transport crash data
  include_client_id: true
  notification_emails:
    - crash@example.com
  bugs:
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
  data_reviews:
    - https://bugzilla.mozilla.org/show_bug.cgi?id=1582479
```

Before we can use our newly defined metric or ping, we need to build the application.  This will cause the [glean_parser](https://github.com/mozilla/glean_parser/)
to work its magic and generate the API files that represent the metrics we defined.

> **NOTE:** Just a little advice from my personal experience here, but I would recommend running the clean task on your project

Reviewer 2: IMO this sounds a bit weird in our book (it would be fine in a blog post!)
maybe change that to be a bit more formal:

> **Note**: Currently you should run the `clean` task on your project after every change to ... That makes sure changes are picked up and that the glean_parser regenerates the code.

or similar

Reviewer 2: Plus, I still have hopes that we will be able to fix this after we have the plugin and better input/output specification

object that was created by the magic of Glean.  You can also see where we are passing in the extra information for the cause
and the message which will get packaged up and sent along with our ping when the second action of `Pings.crash.send()` is
called.  Basically this forces our ping to be sent immediately after the recording of the event.

Finally, we need to register our `MainActivity` as the default uncaught exception handler by adding the following line to the
`onCreate()` function, after `Glean.initialize(this)`:

```Kotlin
Thread.setDefaultUncaughtExceptionHandler(this)
```

### Make A Way To Crash The App

That's pretty much it!  Now we just need to make a way to crash our app with an uncaught exception.  The "Basic Activity"
template has a floating action button called `fab` that can be used, and it's set up in the `onCreate()` of the `MainActivity`.  
We can change the click listener to make it look like this in order to make it throw an exception:

```Kotlin
fab.setOnClickListener {
    // Go boom!
    throw NullPointerException()
}
```

Now run the application and click the floating action button and you should have just crashed the app while recording and

> sending crash telemetry!  I should mention, this information didn't really get recorded by anything, as there is additional

Reviewer 2: As part of our documentation I think it would be better to point this out more formally.
Something along the lines of:

In order for the data to be properly recorded by our ingestion pipeline, additional work is required (link to it, I think we have that documented somewhere?)
This might be more automatic in the future.
